ShaiRandom/Wrappers/ReversingWrapper.cs

```diff
@@ -12,32 +11,47 @@ namespace ShaiRandom.Wrappers
     /// ReversingWrapper will permit calls to its Skip, and they will also go in reverse. The most common use for this would be to run
     /// the wrapped generator forward and track the number of calls made, then run the ReversingWrapper by the same number of calls
     /// (potentially using Skip(), if supported) to revert the state to its original value. This is more convenient for the usage where
-    /// you use <see cref="EnhancedRandomExtensions.Shuffle{T}(IEnhancedRandom, T[])"/> with both the wrapped generator and wrapper, since they will use the same amount of
-    /// calls, and this will even un-shuffle the array (restoring it to its order before the shuffle).
+    /// you use <see cref="EnhancedRandomExtensions.Shuffle{T}(IEnhancedRandom, Span&lt;T&gt;)"/> or other Shuffle variants with both the wrapped
+    /// generator and wrapper, since they will use the same amount of calls, and this will even un-shuffle the array (restoring it to its order before the shuffle).
     /// </summary>
     [Serializable]
-    public class ReversingWrapper : AbstractRandom, IEquatable<ReversingWrapper?>
+    public class ReversingWrapper : AbstractRandom
     {
         /// <summary>
         /// The identifying tag here is "R" , which is an invalid length to indicate the tag is not meant to be registered or used on its own.
         /// </summary>
         public override string Tag => "R";
+
+        /// <summary>
+        /// The ShaiRandom generator being wrapped, which must never be null.
+        /// </summary>
         public IEnhancedRandom Wrapped { get; set; }
 
+        /// <summary>
+        /// Creates a new ReversingWrapper around a new <see cref="FourWheelRandom"/> generator with a random initial
+        /// state.
+        /// </summary>
         public ReversingWrapper()
         {
             Wrapped = new FourWheelRandom();
         }
 
+        /// <summary>
+        /// Creates a new ReversingWrapper around a new <see cref="FourWheelRandom"/> generator seeded with the given
+        /// value.
+        /// </summary>
         public ReversingWrapper(ulong seed)
         {
             Wrapped = new FourWheelRandom(seed);
         }
 
+        /// <summary>
+        /// Creates a new ReversingWrapper around the given generator, which must support <see cref="IEnhancedRandom.PreviousULong"/>.
+        /// </summary>
         public ReversingWrapper(IEnhancedRandom wrapping)
         {
             if (!wrapping.SupportsPrevious)
-                throw new NotSupportedException($"The AbstractRandom to wrap must support PreviousULong(), and {nameof(wrapping)} does not.");
+                throw new ArgumentException($"The AbstractRandom to wrap must support PreviousULong(), and {nameof(wrapping)} does not.", nameof(wrapping));
```

Comment:
I don't actually know what changed here, do you need to specify the interpolated argument twice?

Reply:
The main thing that was changed was the type of exception (now `ArgumentException` instead of `NotSupportedException`); if you look at the documentation for both of those built-in exception types, `ArgumentException` fits the use case more-so than `NotSupportedException` does:

`ArgumentException`: The exception that is thrown when one of the arguments provided to a method is not valid.

`NotSupportedException`: The exception that is thrown when an invoked method is not supported, or when there is an attempt to read, seek, or write to a stream that does not support the invoked functionality.

`ArgumentException` has a few constructors.  One of them just takes a message, and one of them takes both a message, and the name of the parameter that was invalid (which is the one this code is using).  `ArgumentException` has a specific property called `ParamName` which will contain the name of the argument that is passed in as the second parameter, and it will also append this name to the exception message automatically. For example, the `Message` field of `throw new ArgumentException("Argument must be greater than 0", nameof(myParam));` will actually be something like `"Argument must be greater than 0 (Parameter 'myParam')"`.

For this reason, it's probably not necessary to specify it twice, the message could be changed to something like: "The AbstractRandom to wrap must support PreviousULong()".

Reply:
Fixed in most recent commits